### PR TITLE
[flang][driver] Add support for -save-temps 

### PR DESCRIPTION
**Pull request by banach-space**
https://github.com/flang-compiler/f18-llvm-project/pull/1434 was merged and then reverted due to CI failure. Re-opening here with one extra commit which should make the CI happy:  https://github.com/flang-compiler/f18-llvm-project/commit/d43181e452f5df0bf83b644946f16f2785f92303

### CODE REVIEW

File: flang/lib/Frontend/FrontendActions.cpp [hunk 1]
```diff
@@ -66,6 +69,15 @@ bool PrescanAndSemaDebugAction::BeginSourceFileAction() {
 }
 
 bool CodeGenAction::BeginSourceFileAction() {
+  llvmCtx_ = std::make_unique<llvm::LLVMContext>();
+  if (this->currentInput().kind().GetLanguage() == Language::LLVM_IR) {
+    // Parse the bitcode...
+    SMDiagnostic err;
+    llvmModule_ = parseIRFile(currentInput().file(), err, *llvmCtx_);
+
+    return (nullptr != llvmModule_);
```
Comment (Leporacanthicus) [inline flang/lib/Frontend/FrontendActions.cpp:78]: `return !llvmModule_`? I don't think we need the Yoda-style or explicitness? See line 506 and 535

**Review by psteinfeld — COMMENTED** [review]
I don't really understand the content, but I can confirm that all builds and tests correctly and looks good.

File: flang/test/Driver/save-temps.f90 [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+! CWD-NEXT: "-o" "a.out"
+
+!--------------------------
+! `-save-temps=obj`
+!--------------------------
+! Check that temp files are saved in the same directory as the output file
+! regardless of whether -o is specified.
+
+! RUN: %flang -save-temps=obj -fno-integrated-as -o obj/dir/a.out %s -### 2>&1 | FileCheck %s -check-prefix=CHECK-OBJ
+! CHECK-OBJ: "-o" "obj/dir{{/|\\\\}}save-temps.i"
```
Comment (schweitzpgi) [inline flang/test/Driver/save-temps.f90:33]: Why does these checks test for `/` as the first directory separator and either `/` or `\\` as the second directory separator?

Comment (banach-space) [reply flang/test/Driver/save-temps.f90:33]: TBH, I've just copied this [from Clang](https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/save-temps.c#L66-L73) and forgot to check. I couldn't find any rationale in Clang, so I will just change this to plain "/".

File: flang/lib/Frontend/FrontendActions.cpp (hunk 1 above)
Comment (banach-space) [reply flang/lib/Frontend/FrontendActions.cpp:78]: So, we discussed this [before](https://github.com/flang-compiler/f18-llvm-project/pull/1422#discussion_r796839625) and decided that this is a good compromise between being compact and making the intent clear. I hope that that's fine :)

Btw: This change was already reviewed, accepted and merged: https://github.com/flang-compiler/f18-llvm-project/pull/1434

Comment (schweitzpgi) [reply flang/lib/Frontend/FrontendActions.cpp:78]: This should be consistent though and `!llvmModule_` is used elsewhere in this patch. Also, the Yoda style turns up 0 times in MLIR and 3 times in LLVM (and 1 comment), so it appears to be less than popular.

@Leporacanthicus Can you put together a phabricator to make this change in llvm-project? Thanks.

Comment (banach-space) [reply flang/lib/Frontend/FrontendActions.cpp:78]: Just to clarify, I'm not married to the "Yoda style" and don't mind changing it. I use it out of habit - it's muscle memory. I used to work on a project where it was quite pervasive and to me it feels natural. But I agree that we should use a style that's more common and familiar. I do prefer `return (llvmModule_ != nullptr);` over `return !llvmModule_;` though.

> @Leporacanthicus Can you put together a phabricator to make this change in llvm-project? Thanks.

This code is not available in llvm-project/main :) I can fix it myself. In fact, I intend to refactor the driver to use the MLIR style (unless there are strong objections) and can make such tweaks at the same time. But I would like to sync-up `fir-dev` and `main` first (I mean the driver patches).

Comment (sscalpone) [reply flang/lib/Frontend/FrontendActions.cpp:78]: Well defined pointer-to-bool conversion is.  Use it!

**Pull request merged**